app/controllers/api/v0/comments_controller.rb

```diff
@@ -3,21 +3,43 @@ module V0
     class CommentsController < ApiController
       respond_to :json
 
-      caches_action :index,
-                    cache_path: proc { |c| c.params.permit! },
-                    expires_in: 10.minutes
-
-      caches_action :show,
-                    cache_path: proc { |c| c.params.permit! },
-                    expires_in: 10.minutes
+      before_action :set_cache_control_headers, only: %i[index show]
 
       def index
         article = Article.find(params[:a_id])
-        @comments = article.comments.roots.includes(:user).select(%i[id processed_html user_id ancestry])
+
+        @comments = article.comments.includes(:user).select(%i[id processed_html user_id ancestry]).arrange
+
+        set_surrogate_key_header "api_comments_index", edge_cache_keys(@comments)
       end
 
       def show
-        @comment = Comment.find(params[:id].to_i(26))
+        tree_with_root_comment = Comment.subtree_of(params[:id].to_i(26)).
+          includes(:user).
+          select(%i[id processed_html user_id ancestry]).
+          arrange
+
+        # being only one tree we know that the root comment is the first (and only) key
+        @comment = tree_with_root_comment.keys.first
+        @comments = tree_with_root_comment[@comment]
```

Comment:
here we use `ancestry` to have the requested comment and its subtree in separate objects, so we can use the same serializer we use for the index:

```text
- comment 1
  - reply to comment 1
    - reply to reply to comment 1
```

The whole tree will then be navigated by the `edge_cache_keys` method and the JSON serializer